src/Common/src/System/Net/Sockets/SocketAPMExtensions.cs

```diff
@@ -829,6 +829,24 @@ public static int EndSendTo(this Socket socket, IAsyncResult asyncResult)
         {
             return TaskToApm.End<int>(asyncResult);
         }
+
+        // Behavior adapter.
+        private static ArraySegment<byte> CreateArraySegment(byte[] buffer, int offset, int size)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException("buffer");
+            }
+
+            try
+            {
+                return new ArraySegment<byte>(buffer, offset, size);
+            }
+            catch (ArgumentException ae)
+            {
+                throw new ArgumentOutOfRangeException(ae.Message, ae);
```

Comment:
> You're losing the ParamName value in this translation

While technically true, at least with the current ArraySegment implementation, there is no param name supplied when it's creating an ArgumentException, which it only does when array.Length - offset < count.